keps/sig-scheduling/20200630-cost-based-scaling-down-of-pods.md

```diff
@@ -0,0 +1,206 @@
+
+The component will allow for each targeted set of pods to select a different list of scheduling
+constraints. Each pod in a set will be given a cost based on how much important it is in the set.
+The constraints can follow the same rules as the scheduler (through importing scheduling plugins)
+or be custom made (e.g. wrt. to application or proprietary requirements).
+The component will implement a mechanism for ranking pods.
+Either by annotating a pod, updating its status or setting a new field in pod's spec.
+Each controller will have a choice to either ignore the cost or take it into account
+when scaling down.
+
+This way, the logic for selecting a victim for the scaling down operation will be
+separated from each controller. Allowing each consumer to provide its own
+logic for assigning costs. Yet, having all controllers to consume the cost uniformly.
+
+Given the default scheduler is not a source of truth about how a pod should be distributed
+after it was scheduled, scaling down strategies can exercise completely different approaches.
+
+Examples of scheduling constraints:
+- choose pods running on a node which have a `PreferNoSchedule` taint first
+- choose youngest/oldest pods first
+- choose pods minimizing topology skew among failure domains (e.g. availability zones)
+
+The goal of the proposal is not to provide specific strategies for more informed scaling down operation.
+The primary goal is to provide a mechanism and have consumers of the new component to define their own
+strategies.
+
+### User Stories [optional]
+
+#### Story 1
+
+From [@pnovotnak](https://github.com/kubernetes/kubernetes/issues/4301#issuecomment-328685358):
+
+```
+I have a number of scientific programs that I've wrapped with code to talk
+to a message broker that do not checkpoint state. The cost of deleting the resource
+increases over time (some of these tasks take hours), until it completes the current unit of work.
+
+Choosing a pod by most idle resources would also work in my case.
+```
+
+#### Story 2
+
+From [@cpwood](https://github.com/kubernetes/kubernetes/issues/4301#issuecomment-436587548)
+
+```
+For my use case, I'd prefer Kubernetes to choose its victims from pods that are running on nodes which have a PreferNoSchedule taint.
+```
+
+### Implementation Details/Notes/Constraints
+
+Currently, the descheduler does not allow to immediately react on changes in a cluster.
+Yet, with some modification another instance of the descheduler (with different set of strategies)
+might be ran in watch mode and rank each pods as it comes.
+Also, once the scheduling framework gets migrated into its own repository,
+scheduling plugins can be vendored as well to provide some of the core scheduling logic.
+
+Once a pod is removed, ranks of all relevant pods are re-computed.
+The pod ranking is best-effort so in case a controller is to delete more than one pod
```

Comment:
IMO "Delete more than one pod" would be a common req, "selects all the pods with the smallest cost and remove" may flip and imbalance situation easily. So is it possible to ensure the "cost" design to support "Delete X pods in a batch"? No need to be perfect in the beginning, but keep the batch-deletion req in mind from day 1 is necessary.

Reply:
I have added a few words on this topic. There are two categories of strategies defined. One category where there's no need to recompute costs. So it's safe to scale down with 2 or more replicas at the same time. Other category where each time a pod is added/removed, all the costs need to be re-computed. I am mainly focusing on the first category to provide basic functionality. With a space to discuss the second category.